server/src/main/java/org/opensearch/action/search/SearchRequestOperationsListeners.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * SearchRequestOperationsListeners contains listeners registered to search requests,
+ * and is responsible for creating the {@link SearchRequestOperationsListener.CompositeListener}
+ * with the all listeners enabled at cluster-level and request-level.
+ *
+ *
+ * @opensearch.internal
+ */
+public class SearchRequestOperationsListeners {
```

Comment:
I don't really understand the use of this at this point. It is nothing but essentially what Composite listener is giving you. Why can't we use that instead?

Reply:
the Composite listener instance is created on every request, while this `SearchRequestOperationsListeners` is more of a group of "global" listeners that exists on the node level, it contains all listeners that are registered, and is responsible for constructing the Composite listener instance based on the `enable` state of each individual listener. 

This is a needed prerequisite for other plugins (like query insight) to use the `SearchRequestOperationsListener`, also it would make the search path cleaner since it would be a bad practice to add more and more listeners in the TransportSearchAction class like this: https://github.com/opensearch-project/OpenSearch/blob/0425c0794ee58ec1af0d41f5294fd4237829f69a/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L190

Reply:
> Also the Composite listener instance is created on every request, while this SearchRequestOperationsListeners is more of a group of "global" listeners that exists on the node level, it contains all listeners that are registered, and is responsible for constructing the Composite listener instance based on the enable state of each individual listener.

Right -- this one is essentially a singleton "CompositeListenerFactory". The underlying listeners can be enabled/disabled based on cluster settings (or theoretically some other state change). For each request, `SearchRequestOperationsListeners` will create an immutable `CompositeListener` with the set of listeners that were enabled at the start of the request.

Reply:
Maybe better to update the class name to SearchRequestOperationsCompositeListenerFactory? Makes it much more clear instead of subtle s between the 2 class. Also class can be made final to prevent unexpected subclasses for now